src/checkers/inference/solver/util/PrintUtils.java

```diff
@@ -159,146 +151,212 @@ private static void buildStatisticText(String key, Integer value, StringBuilder
         statisticsText.append("\n");
     }
 
-    public static void printUnsolvable(Collection<Constraint> unsatisfactoryConstraints) {
-        if (unsatisfactoryConstraints == null || unsatisfactoryConstraints.isEmpty()) {
-            System.out.println("The backend you used doesn't support explanation feature!");
-            return;
-        }
 
+    /**
+     * Outputs the statistics to the given stream.
+     * @param stream
+     * @param statistic
+     * @param modelRecord
+     */
+    private static void outputStatistics(PrintStream stream, Map<StatisticKey, Long> statistic,
+            Map<String, Integer> modelRecord) {
+        stream.println("/***********************Statistics*************************/");
+        stream.println(buildStatistic(statistic, modelRecord));
+        stream.println("/**********************************************************/");
+    }
+
+    /**
+     * Print the statistics to screen.
+     */
+    public static void printStatistics(Map<StatisticKey, Long> statistics,
+            Map<String, Integer> modelRecord) {
+        outputStatistics(System.out, statistics, modelRecord);
+    }
+
+    /**
+     * Write the statistics to a file called statistics.txt.
+     *
+     * @param statistics a map between stats keys and their long values.
+     * @param modelRecord
+     * @param appendWrite if set to true, will append the solutions to the output file.
+     */
+    public static void writeStatistics(Map<StatisticKey, Long> statistics,
+            Map<String, Integer> modelRecord,
+            boolean appendWrite) {
+        File outFile = new File("statistics.txt");
+        PrintStream out = getFilePrintStream(outFile, appendWrite);
+        outputStatistics(out, statistics, modelRecord);
+        out.close();
+        System.out.println("Statistics have been written to: " + outFile.getAbsolutePath() + "\n");
+    }
+
+
+    private static String generateUnsolveableString(Collection<Constraint> unsatisfactoryConstraints) {
         ToStringSerializer toStringSerializer = new ToStringSerializer(false);
         SlotPrinter slotPrinter = new SlotPrinter(toStringSerializer);
+
+        StringBuffer sb = new StringBuffer();
```

Comment:
rework this method to pass down the string buffer, update toStringSerializer and slotPrinter